### DISCUSSION TITLE
Annotated NewType Path parameter with regex constraint throws error

### DISCUSSION THREAD

**Discussion by dozed** [fastapi/fastapi#11054] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from __future__ import annotations

from typing import NewType, Annotated

import uvicorn
from fastapi import FastAPI, Path
from starlette.responses import Response

app = FastAPI()

BookId = NewType('BookId', str)

BookIdParam = Annotated[BookId, Path(pattern='^(?!.*\\.\\.)[a-zA-Z0-9-_]+$')]


@app.get('/books/{book_id}')
async def get_recursive_dataclass(book_id: BookIdParam) -> Response:
    return Response(status_code=200, content='ok')


if __name__ == '__main__':
    uvicorn.run('main:app', reload=True)
```

### Description

Running the code gives the following error:

```
ValueError: On field "book_id" the following field constraints are set but not enforced: regex.
For more details see https://docs.pydantic.dev/usage/schema/#unenforced-field-constraints
```

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

1.10.8

### Python Version

3.11.7

**Comment by fmelihh** [comment, marked as answer]
Hi @dozed Here is the solution

```
from __future__ import annotations

from typing import NewType

import uvicorn
from fastapi import FastAPI
from starlette.responses import Response
from pydantic import constr

app = FastAPI()

BookId = NewType('BookId', constr(regex='^(?!.*\\.\\.)[a-zA-Z0-9-_]+$'))

BookIdParam = BookId


@app.get('/books/{book_id}')
async def get_recursive_dataclass(book_id: BookIdParam) -> Response:
    return Response(status_code=200, content='ok')


if __name__ == '__main__':
    uvicorn.run('main:app', reload=True)
```

The issue you are facing is related to using the NewType for BookId and the Annotated type hint for BookIdParam. The NewType is essentially creating a new type that is identical to the underlying type (in this case, str). When you use NewType to create a new type, it doesn't carry over any annotations or constraints from the original type.
Happy coding :)

**Reply by dozed** [reply]
Thanks, that helps :)

I would slightly prefer to not have the constraints on the actual domain types and rather have them annotated at the API layer (e.g. parameter definitions), but I guess I can live with it ;)
